relay-server/src/actors/project_cache.rs

```diff
@@ -468,7 +530,41 @@ impl ProjectCacheBroker {
         } = message;
 
         self.get_or_create_project(project_key)
-            .update_state(state, no_cache)
+            .update_state(state.clone(), no_cache);
+
+        // Envelopes need to remain in the queue while Relay receives invalid states from upstream.
+        if state.invalid() {
+            return;
+        }
+
+        let envelopes = self.pending_envelopes.dequeue(&project_key, |queue_key| {
+            // Pick envelopes which belong to the incoming project.
+            if project_key != queue_key.key {
+                return false;
+            }
+
+            // We return false if project is not cached or its state is invalid.
+            self.projects
+                .get(&queue_key.sampling_key)
+                // Make sure we have only cached and valid state.
+                .and_then(|p| p.valid_state())
+                .map_or(false, |s| !s.invalid())
+        });
+
+        // Flush envelopes where both states have resolved.
+        for (envelope, envelope_context) in envelopes {
+            let sampling_state = utils::get_sampling_key(&envelope)
+                .and_then(|key| self.projects.get(&key))
+                .and_then(|p| p.valid_state());
+
+            self.handle_processing(
+                project_key,
```

Comment:
With the suggestion I made above, `handle_processing` should take the envelope's project key, not this one.